src/Microsoft.Restier.Core/ApiConfiguration.cs

```diff
@@ -65,5 +68,46 @@ public IServiceProvider ServiceProvider
         {
             return this.serviceProvider.GetService<T>();
         }
+
+        internal TaskCompletionSource<IEdmModel> CompeteModelGeneration(out Task<IEdmModel> running)
+        {
+            var source = new TaskCompletionSource<IEdmModel>(TaskCreationOptions.AttachedToParent);
+            var exist = Interlocked.CompareExchange(ref modelTask, source.Task, null);
+            if (exist != null)
+            {
+                running = exist;
+                source.SetCanceled();
+                return null;
+            }
+
+            source.Task.ContinueWith(
+                task => FinishModelGeneration(task),
+                TaskContinuationOptions.ExecuteSynchronously);
+            running = null;
+            return source;
+        }
+
+        private void FinishModelGeneration(Task<IEdmModel> task)
+        {
+            if (task.Status == TaskStatus.RanToCompletion)
+            {
+                ReplaceModelTask(Task.FromResult(task.Result), task);
```

Comment:
Intuitively I think a `Task` instance primarily contains the id information, state variable, the pointer to a task code. So I would assume that all task instances are of the same size. So it's OK to leave the completed task as is. But I do think the design to replace the `modelTask` with `null` to allow retry is very good:)